### PR TITLE
NF: add: Support updating a previous container

### LINKED ISSUE #58: Possibility to overwrite/update an existing container?
I had already a container for the version 1.0 and decided to upgrade to 1.1:
```shell
adina@head1:/data/movieloc/backup_store/saccs$ datalad containers-list
containers(ok): /data/movieloc/backup_store/saccs/.datalad/environments/BIDSSacc/image (file)
containers(ok): /data/movieloc/backup_store/saccs/.datalad/environments/fsl/image (file)
action summary:
  containers (ok: 2)
adina@head1:/data/movieloc/backup_store/saccs$ datalad containers-add --url shub://AdinaWagner/BIDSsacc:1.1 BIDSSacc
[WARNING] Running addurl resulted in stderr output:   while adding a new url to an already annexed file, url does not have expected file size (use --relaxed to bypass this check) https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2FAdinaWagner%2FBIDSsacc%2F0d4acf64f18fb59b91e99a7eba785b986f5d60b9%2F2a6fd9f0256615a8151b55ea00e8524d%2F2a6fd9f0256615a8151b55ea00e8524d.simg?generation=1543600024406021&alt=media
git-annex: addurl: 1 failed
 
[ERROR  ] CommandError: command 'addurl'
| Error, annex reported failure for addurl (url='https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2FAdinaWagner%2FBIDSsacc%2F0d4acf64f18fb59b91e99a7eba785b986f5d60b9%2F2a6fd9f0256615a8151b55ea00e8524d%2F2a6fd9f0256615a8151b55ea00e8524d.simg?generation=1543600024406021&alt=media'): {'command': 'addurl', 'file': '.datalad/environments/BIDSSacc/image', 'success': False} [containers_add(/data/movieloc/backup_store/saccs/.datalad/environments/BIDSSacc/image)] 
containers_add(error): /data/movieloc/backup_store/saccs/.datalad/environments/BIDSSacc/image (file) [CommandError: command 'addurl'
Error, annex reported failure for addurl (url='https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2FAdinaWagner%2FBIDSsacc%2F0d4acf64f18fb59b91e99a7eba785b986f5d60b9%2F2a6fd9f0256615a8151b55ea00e8524d%2F2a6fd9f0256615a8151b55ea00e8524d.simg?generation=1543600024406021&alt=media'): {'command': 'addurl', 'file': '.datalad/environments/BIDSSacc/image', 'success': False}]
containers_add(ok): /data/movieloc/backup_store/saccs/.datalad/environments/BIDSSacc/image (file) [CommandError: command 'addurl'
Error, annex reported failure for addurl (url='https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2FAdinaWagner%2FBIDSsacc%2F0d4acf64f18fb59b91e99a7eba785b986f5d60b9%2F2a6fd9f0256615a8151b55ea00e8524d%2F2a6fd9f0256615a8151b55ea00e8524d.simg?generation=1543600024406021&alt=media'): {'command': 'addurl', 'file': '.datalad/environments/BIDSSacc/image', 'success': False}]
action summary:
  containers_add (error: 1, ok: 1)
  save (notneeded: 1)
```
I was initially "pleased" that containers-add didn't just say "cannot do it since you already have a container with such name", but the reason of failure was not obvious. Apparently an attempt to provide a new image for an already registered container would fail if remote file changed its size. The reason becomes obvious if we look at non`--json` invocation of `addurl` where more information is provided (@joeyh - would be nice to have it also reported in the msg of the json record):

```shell
adina@head1:/data/movieloc/backup_store/saccs$ 'git' '-c' 'receive.autogc=0' '-c' 'gc.auto=0' 'annex' 'addurl' '--json' '--json-progress' '--file=.datalad/environments/BIDSSacc/image' 'https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2FAdinaWagner%2FBIDSsacc%2F0d4acf64f18fb59b91e99a7eba785b986f5d60b9%2F2a6fd9f0256615a8151b55ea00e8524d%2F2a6fd9f0256615a8151b55ea00e8524d.simg?generation=1543600024406021&alt=media' '--'
  while adding a new url to an already annexed file, url does not have expected file size (use --relaxed to bypass this check) https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2FAdinaWagner%2FBIDSsacc%2F0d4acf64f18fb59b91e99a7eba785b986f5d60b9%2F2a6fd9f0256615a8151b55ea00e8524d%2F2a6fd9f0256615a8151b55ea00e8524d.simg?generation=1543600024406021&alt=media
{"command":"addurl","success":false,"file":".datalad/environments/BIDSSacc/image"}
git-annex: addurl: 1 failed
adina@head1:/data/movieloc/backup_store/saccs$ 'git' '-c' 'receive.autogc=0' '-c' 'gc.auto=0' 'annex' 'addurl' '--file=.datalad/environments/BIDSSacc/image' 'https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2FAdinaWagner%2FBIDSsacc%2F0d4acf64f18fb59b91e99a7eba785b986f5d60b9%2F2a6fd9f0256615a8151b55ea00e8524d%2F2a6fd9f0256615a8151b55ea00e8524d.simg?generation=1543600024406021&alt=media' '--'
addurl https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2FAdinaWagner%2FBIDSsacc%2F0d4acf64f18fb59b91e99a7eba785b986f5d60b9%2F2a6fd9f0256615a8151b55ea00e8524d%2F2a6fd9f0256615a8151b55ea00e8524d.simg?generation=1543600024406021&alt=media 
  while adding a new url to an already annexed file, url does not have expected file size (use --relaxed to bypass this check) https://www.googleapis.com/download/storage/v1/b/singularityhub/o/singularityhub%2Fgithub.com%2FAdinaWagner%2FBIDSsacc%2F0d4acf64f18fb59b91e99a7eba785b986f5d60b9%2F2a6fd9f0256615a8151b55ea00e8524d%2F2a6fd9f0256615a8151b55ea00e8524d.simg?generation=1543600024406021&alt=media
failed
```

datalad-containers version 0.2.1

### PR DESCRIPTION
**Pull request by kyleam**
Add `--update` switch to `containers-add` to make it possible to update aspects of an existing container.

Closes #58.

### CODE REVIEW

**Comment by kyleam** [comment]
Pushed an additional commit to deal with #37.

**Comment by kyleam** [comment]
> Pushed an additional commit to deal with #37.

Ooops, and broke something else.  Will have a look.

**Comment by bpoldrack** [comment]
So, I guess this is ready to be merged or do you intend to add something?

File: datalad_container/containers_add.py [hunk 1]
```diff
@@ -170,6 +204,14 @@ def __call__(name, url=None, dataset=None, call_fmt=None, image=None):
         to_save = []
         imgurl = url
         if url:
+            if update and op.lexists(image):
+                # XXX: check=False is used to avoid dropping the image. It
+                # should use drop=False if remove() gets such an option (see
+                # DataLad's gh-2673).
+                for r in ds.remove(image, save=False, check=False,
```
Comment (yarikoptic) [inline datalad_container/containers_add.py]: Hm, iirc remove also drops that file content, so we might loose that previous version of image forever. I would have just used regular gitrepo remove for now here

Comment (kyleam) [reply datalad_container/containers_add.py]: > Hm, iirc remove also drops that file content

I don't think that is true. 

```
% datalad run "echo blah >blah"
% datalad remove --nocheck blah
remove(ok): blah
save(ok): /tmp/tmp.Vn2kcJWMVw-dl (dataset)
action summary:
  remove (ok: 1)
  save (ok: 1)
% git-annex findref master~
blah
```

**Comment by kyleam** [comment]
> So, I guess this is ready to be merged or do you intend to add something?

No, didn't plan to add anything else unless needed to address objections :]

File: datalad_container/containers_add.py (hunk 1 above)
Comment (yarikoptic) [reply datalad_container/containers_add.py]: hm, indeed... https://github.com/datalad/datalad/issues/2673#issuecomment-400686307  - so not dropped if `--nocheck`.  really obscure but in effect correct here, so could be accepted as is I guess for now. Thanks

Comment (kyleam) [reply datalad_container/containers_add.py]: To be fair, this change came with a comment that points datalad/datalad#2673 and mentions the reason for the weirdness (it's even visible from your inline comment).

**Pull request merged**